File: runtime/lib/phases/fetch.ts
import { createReadStream, createWriteStream } from 'node:fs';
import { mkdir, rename, rm, writeFile } from 'node:fs/promises';
import { dirname } from 'node:path';
import { Readable, Transform } from 'node:stream';
import type { ReadableStream as NodeReadableStream } from 'node:stream/web';
import { pipeline } from 'node:stream/promises';
import type { Artifact } from '@torba/core';
import {
  isSourceUrl,
  isSourceFile,
  isSourceString,
  isSourceBytes,
  interpolate,
} from '@torba/core';

export interface FetchTask {
  artifact: Artifact;
  finalPath: string;
}

export interface FetchHooks {
  onStart?: (task: FetchTask) => void;
  onBytes?: (task: FetchTask, bytes: number) => void;
  onDone?: (task: FetchTask) => void;
}

const RETRY_DELAYS_MS = [500, 2_000, 8_000];

function counter(onBytes: (n: number) => void): Transform {
  let total = 0;
  return new Transform({
    transform(chunk, _enc, cb) {
      total += chunk.length;
      onBytes(total);
      cb(null, chunk);
    },
  });
}

async function fetchOnce(
  task: FetchTask,
  vars: Record<string, string>,
  onBytes: (n: number) => void,
): Promise<void> {
  const { artifact, finalPath } = task;
  const src = artifact.source;
  await mkdir(dirname(finalPath), { recursive: true });
  const tmpPath = `${finalPath}.partial`;

  if (isSourceUrl(src)) {
    const url = interpolate(src.url, vars);
    const res = await fetch(url);
    if (!res.ok) throw new Error(`HTTP ${res.status} for ${url}`);
    if (res.body) {
      await pipeline(
        Readable.fromWeb(res.body as unknown as NodeReadableStream),
        counter(onBytes),
        createWriteStream(tmpPath),
      );
    } else {
      await writeFile(tmpPath, '');
    }
  } else if (isSourceFile(src)) {
    await pipeline(
      createReadStream(interpolate(src.file, vars)),
      counter(onBytes),
      createWriteStream(tmpPath),
    );
  } else if (isSourceString(src)) {
    await writeFile(tmpPath, src.string);
    onBytes(Buffer.byteLength(src.string));
  } else if (isSourceBytes(src)) {
    const buf = Buffer.from(src.bytes, 'base64');
    await writeFile(tmpPath, buf);
    onBytes(buf.length);
  } else {
    throw new Error(`Unsupported source for ${artifact.path}`);
  }

  await rename(tmpPath, finalPath);
}

async function fetchOne(
  task: FetchTask,
  vars: Record<string, string>,
  onBytes: (n: number) => void,
): Promise<void> {
  for (let attempt = 0; ; attempt++) {
    try {
      await fetchOnce(task, vars, onBytes);
      return;
    } catch (err) {
      await rm(`${task.finalPath}.partial`, { force: true });
      if (attempt >= RETRY_DELAYS_MS.length) throw err;
      await new Promise((r) => setTimeout(r, RETRY_DELAYS_MS[attempt]));
    }
  }
}

/**
 * Concurrency cost of a download by size, against a budget that defaults to 8.
 * The buckets give: many tiny files run 8-wide, medium files 4-wide, large
 * 2-wide, and one really big file gets the pipe to itself — so a single
 * fat jar isn't split eight ways. Unknown size is treated as small.
 */
function weight(size: number | undefined): number {
  if (size === undefined) return 1;
  const MB = 1024 * 1024;
  if (size < 1 * MB) return 1; // small
  if (size < 10 * MB) return 2; // medium
  if (size < 50 * MB) return 4; // large
  return 8; // huge — runs alone at the default budget
}

/**
 * Weighted semaphore: admits any waiter whose weight currently fits, in queue
 * order. Skipping a too-big head to admit a smaller follower avoids
 * head-of-line blocking when an LPT-sorted queue puts the heaviest task first.
 */
class Budget {
  private used = 0;
  private waiters: Array<{ need: number; resolve: () => void }> = [];
  constructor(private readonly cap: number) {}

  async acquire(amount: number): Promise<void> {
    const need = Math.min(amount, this.cap);
    if (this.used + need <= this.cap) {
      this.used += need;
      return;
    }
    return new Promise((resolve) => {
      this.waiters.push({ need, resolve });
    });
  }

  release(amount: number): void {
    const need = Math.min(amount, this.cap);
    this.used -= need;
    for (let i = 0; i < this.waiters.length; ) {
      const w = this.waiters[i]!;
      if (this.used + w.need <= this.cap) {
        this.waiters.splice(i, 1);
        this.used += w.need;
        w.resolve();
      } else {
        i++;
      }
    }
  }
}

export async function fetchAll(
  tasks: FetchTask[],
  vars: Record<string, string>,
  concurrency: number,
  hooks: FetchHooks = {},
): Promise<void> {
  // Largest-first so the heavy files claim budget early and smaller files
  // backfill as bytes free up. Unknown size sinks to the end.
  const ordered = [...tasks].sort(
    (a, b) => (b.artifact.size ?? 0) - (a.artifact.size ?? 0),
  );

  const budget = new Budget(concurrency);
  await Promise.all(
    ordered.map(async (task) => {
      const w = weight(task.artifact.size);
      await budget.acquire(w);
      try {
        hooks.onStart?.(task);
        await fetchOne(task, vars, (n) => hooks.onBytes?.(task, n));
        hooks.onDone?.(task);
      } finally {
        budget.release(w);
      }
    }),
  );
}

File: crates/opys-runtime/src/phases/fetch.rs
use base64::Engine;
use futures::StreamExt;
use indexmap::IndexMap;
use std::path::Path;
use std::sync::Arc;
use tokio::fs;
use tokio::io::AsyncWriteExt;
use tokio::sync::Mutex;
use tokio::task::JoinSet;
use tokio_util::sync::CancellationToken;
use opys_core::{interpolate, Artifact, Source};

use crate::errors::InstallError;
use crate::fetch::{client, OPYS_USER_AGENT};

pub struct FetchTask {
    pub artifact: Artifact,
    pub final_path: String,
}

type FetchStartHook = Arc<dyn Fn(&FetchTask) + Send + Sync>;
type FetchBytesHook = Arc<dyn Fn(&FetchTask, u64) + Send + Sync>;
type FetchDoneHook = Arc<dyn Fn(&FetchTask) + Send + Sync>;

#[derive(Clone, Default)]
pub struct FetchHooks {
    pub on_start: Option<FetchStartHook>,
    pub on_bytes: Option<FetchBytesHook>,
    pub on_done: Option<FetchDoneHook>,
}

const RETRY_DELAYS_MS: &[u64] = &[500, 2_000, 8_000];

async fn fetch_once(
    task: &FetchTask,
    vars: &IndexMap<String, String>,
    on_bytes: &(dyn Fn(u64) + Send + Sync),
) -> Result<(), InstallError> {
    let final_path = Path::new(&task.final_path);
    if let Some(parent) = final_path.parent() {
        fs::create_dir_all(parent).await.map_err(|source| InstallError::Io {
            path: parent.display().to_string(),
            source,
        })?;
    }
    let tmp_path = format!("{}.partial", task.final_path);

    match &task.artifact.source {
        Source::Url { url } => {
            let url = interpolate(url, vars);
            let res = client()
                .get(&url)
                .header("user-agent", OPYS_USER_AGENT)
                .send()
                .await
                .map_err(|e| InstallError::Network {
                    url: url.clone(),
                    status: 0,
                    body: e.to_string(),
                })?;
            let status = res.status();
            if !status.is_success() {
                let body = res.text().await.unwrap_or_default();
                return Err(InstallError::Network {
                    url,
                    status: status.as_u16(),
                    body,
                });
            }
            let mut file = fs::File::create(&tmp_path).await.map_err(|source| {
                InstallError::Io {
                    path: tmp_path.clone(),
                    source,
                }
            })?;
            let mut total: u64 = 0;
            let mut stream = res.bytes_stream();
            while let Some(chunk) = stream.next().await {
                let chunk = chunk.map_err(|e| InstallError::Network {
                    url: url.clone(),
                    status: 0,
                    body: e.to_string(),
                })?;
                file.write_all(&chunk).await.map_err(|source| InstallError::Io {
                    path: tmp_path.clone(),
                    source,
                })?;
                total += chunk.len() as u64;
                on_bytes(total);
            }
            file.flush().await.ok();
        }
        Source::File { file: path } => {
            let resolved = interpolate(path, vars);
            let data = fs::read(&resolved).await.map_err(|source| InstallError::Io {
                path: resolved.clone(),
                source,
            })?;
            let len = data.len() as u64;
            fs::write(&tmp_path, &data).await.map_err(|source| InstallError::Io {
                path: tmp_path.clone(),
                source,
            })?;
            on_bytes(len);
        }
        Source::String { string } => {
            let bytes = string.as_bytes();
            fs::write(&tmp_path, bytes).await.map_err(|source| InstallError::Io {
                path: tmp_path.clone(),
                source,
            })?;
            on_bytes(bytes.len() as u64);
        }
        Source::Bytes { bytes } => {
            let decoded = base64::engine::general_purpose::STANDARD
                .decode(bytes)
                .map_err(|e| InstallError::other(format!("bad base64: {e}")))?;
            let len = decoded.len() as u64;
            fs::write(&tmp_path, &decoded).await.map_err(|source| InstallError::Io {
                path: tmp_path.clone(),
                source,
            })?;
            on_bytes(len);
        }
        Source::Pointer { .. } => {
            // Pointer sources should be resolved before scan/fetch.
            return Err(InstallError::other(format!(
                "Unsupported source for {}",
                task.artifact.path
            )));
        }
    }

    fs::rename(&tmp_path, &task.final_path)
        .await
        .map_err(|source| InstallError::Io {
            path: tmp_path,
            source,
        })?;
    Ok(())
}

async fn fetch_one(
    task: &FetchTask,
    vars: &IndexMap<String, String>,
    on_bytes: &(dyn Fn(u64) + Send + Sync),
) -> Result<(), InstallError> {
    let mut attempt = 0usize;
    loop {
        match fetch_once(task, vars, on_bytes).await {
            Ok(()) => return Ok(()),
            Err(err) => {
                let _ = fs::remove_file(format!("{}.partial", task.final_path)).await;
                if attempt >= RETRY_DELAYS_MS.len() {
                    return Err(err);
                }
                tokio::time::sleep(std::time::Duration::from_millis(RETRY_DELAYS_MS[attempt]))
                    .await;
                attempt += 1;
            }
        }
    }
}

/// Size → concurrency weight. Buckets match the TS version: small=1, med=2,
/// large=4, huge=8 (alone at the default budget).
fn weight(size: Option<u64>) -> u32 {
    let Some(size) = size else { return 1 };
    const MB: u64 = 1024 * 1024;
    if size < MB {
        1
    } else if size < 10 * MB {
        2
    } else if size < 50 * MB {
        4
    } else {
        8
    }
}

/// Weighted-budget semaphore: admits any waiter whose weight currently fits,
/// in arrival order. Skipping a too-big head to admit a smaller follower
/// avoids head-of-line blocking on the LPT-sorted queue.
struct Budget {
    state: Mutex<BudgetState>,
}

struct BudgetState {
    cap: u32,
    used: u32,
    waiters: Vec<Waiter>,
}

struct Waiter {
    need: u32,
    tx: tokio::sync::oneshot::Sender<()>,
}

impl Budget {
    fn new(cap: u32) -> Self {
        Self {
            state: Mutex::new(BudgetState {
                cap: cap.max(1),
                used: 0,
                waiters: Vec::new(),
            }),
        }
    }

    async fn acquire(&self, amount: u32) {
        let rx = {
            let mut state = self.state.lock().await;
            let need = amount.min(state.cap);
            if state.used + need <= state.cap {
                state.used += need;
                return;
            }
            let (tx, rx) = tokio::sync::oneshot::channel();
            state.waiters.push(Waiter { need, tx });
            rx
        };
        rx.await.ok();
    }

    async fn release(&self, amount: u32) {
        let mut state = self.state.lock().await;
        let need = amount.min(state.cap);
        state.used = state.used.saturating_sub(need);
        let mut i = 0;
        while i < state.waiters.len() {
            let cap = state.cap;
            let used = state.used;
            let need = state.waiters[i].need;
            if used + need <= cap {
                let w = state.waiters.remove(i);
                state.used += w.need;
                let _ = w.tx.send(());
            } else {
                i += 1;
            }
        }
    }
}

pub async fn fetch_all(
    mut tasks: Vec<FetchTask>,
    vars: &IndexMap<String, String>,
    concurrency: u32,
    hooks: FetchHooks,
    cancel: &CancellationToken,
) -> Result<(), InstallError> {
    // Largest-first.
    tasks.sort_by(|a, b| b.artifact.size.unwrap_or(0).cmp(&a.artifact.size.unwrap_or(0)));

    let budget = Arc::new(Budget::new(concurrency));
    let vars = Arc::new(vars.clone());

    // A JoinSet (not bare `tokio::spawn`) so that returning early — on error or
    // cancellation — drops the set and aborts every still-running download.
    // Detached `tokio::spawn` tasks would instead keep running in the background.
    let mut set: JoinSet<Result<(), InstallError>> = JoinSet::new();
    for task in tasks {
        let budget = Arc::clone(&budget);
        let vars = Arc::clone(&vars);
        let hooks = hooks.clone();
        let task = Arc::new(task);
        set.spawn(async move {
            let w = weight(task.artifact.size);
            budget.acquire(w).await;
            let res: Result<(), InstallError> = async {
                if let Some(h) = &hooks.on_start {
                    h(&task);
                }
                let t = Arc::clone(&task);
                let hooks_inner = hooks.clone();
                fetch_one(&task, &vars, &|n| {
                    if let Some(h) = &hooks_inner.on_bytes {
                        h(&t, n);
                    }
                })
                .await?;
                if let Some(h) = &hooks.on_done {
                    h(&task);
                }
                Ok(())
            }
            .await;
            budget.release(w).await;
            res
        });
    }

    loop {
        tokio::select! {
            // Bias the cancel check so a pending cancellation always wins over a
            // ready download result.
            biased;
            _ = cancel.cancelled() => return Err(InstallError::Cancelled),
            joined = set.join_next() => match joined {
                None => return Ok(()),
                Some(Ok(Ok(()))) => {}
                Some(Ok(Err(e))) => return Err(e),
                Some(Err(e)) => return Err(InstallError::other(format!("join error: {e}"))),
            },
        }
    }
}
